java/src/main/java/io/cucumber/gherkin/GherkinDocumentBuilder.java

```diff
@@ -245,32 +245,41 @@ private static String joinMatchedText(List<Token> lineTokens) {
         return content.toString();
     }
 
+    @SuppressWarnings("ForLoopReplaceableByForEach") // classic 'for' loop is ~2x faster than 'for-each'
     private List<TableRow> getTableRows(AstNode node) {
         List<Token> tokens = node.getTokens(TokenType.TableRow);
-        List<TableRow> rows = new ArrayList<>(tokens.size());
+        int tokenCount = tokens.size();
+        List<TableRow> rows = new ArrayList<>(tokenCount);
 
-        for (Token token : tokens) {
+        for (int i = 0; i < tokenCount; i++) {
+            Token token = tokens.get(i);
             rows.add(new TableRow(token.location, getCells(token), idGenerator.newId()));
         }
         ensureCellCount(rows);
         return rows;
     }
 
+    @SuppressWarnings("ForLoopReplaceableByForEach") // classic 'for' loop is ~2x faster than 'for-each'
     private void ensureCellCount(List<TableRow> rows) {
         if (rows.isEmpty())
             return;
 
         int cellCount = rows.get(0).getCells().size();
-        for (TableRow row : rows) {
+        for (int i = 0, rowsSize = rows.size(); i < rowsSize; i++) {
+            TableRow row = rows.get(i);
             if (row.getCells().size() != cellCount) {
                 throw new ParserException.AstBuilderException("inconsistent cell count within the table", row.getLocation());
             }
         }
     }
 
+    @SuppressWarnings("ForLoopReplaceableByForEach") // classic 'for' loop is ~2x faster than 'for-each'
```

Comment:
How was this established?  I can't seem to reproduce it.

I've done a very small sample test. 3 samples with and without the optimization. With the optimization enabled I get between 330ms and 380ms for `getCells`. Without the optimization 280ms to 480ms. Variability aside, it's not a 2x difference.

Reply:
The 2x factor was determined using the profile, not through micro-benchmark. With JMH, I would expect to have a lower difference, or no difference at all if the JIT compiler works well. The results when reverting the commit is unexpected (for-each being faster than classical for). That's a behavior that is noticed in https://reneschwietzke.de/java/java-loop-performance.html but only when disabling the JIT compiler.

As we have no evidence of better performance with the classical for, so let's keep the for-each, this makes the code more readable.